cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java

```diff
@@ -152,21 +153,63 @@ public MetadataEntity toMetadataEntity() {
    * metadataEntity.
    */
   public static <T extends EntityId> T fromMetadataEntity(MetadataEntity metadataEntity) {
-    EntityType entityType = EntityType.valueOf(metadataEntity.getType().toUpperCase());
-    if ((entityType == EntityType.APPLICATION || entityType == EntityType.PROGRAM) &&
-      metadataEntity.getValue(MetadataEntity.VERSION) == null) {
-      // if the EntityType is application or program and a version is not specified in the MetadataEntity then add it
-      if (entityType == EntityType.APPLICATION) {
-        metadataEntity = metadataEntity.append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION);
-      } else {
-        metadataEntity = MetadataEntity.ofNamespace(metadataEntity.getValue(MetadataEntity.NAMESPACE))
-          .append(MetadataEntity.APPLICATION, metadataEntity.getValue(MetadataEntity.APPLICATION))
-          .append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION)
-          .append(MetadataEntity.TYPE, metadataEntity.getValue(MetadataEntity.TYPE))
-          .append(MetadataEntity.PROGRAM, metadataEntity.getValue(MetadataEntity.PROGRAM));
+    EntityType.valueOf(metadataEntity.getType().toUpperCase());
+    return getKnownParentEntityId(metadataEntity);
```

Comment:
I think this method is ill-named, because in some cases it does not return the equivalent EntityId, but a parent id. I am not sure where all this is used, but I think it would be better to name it in a way that indicates this behavior. Also, change the Javadocs to explain this.

Reply:
Done.